internal/driver/glfw/window.go

```diff
@@ -430,17 +430,29 @@ func (w *window) Close() {
 		return
 	}
 
-	w.viewLock.Lock()
-	w.closing = true
-	w.viewLock.Unlock()
-	w.viewport.SetShouldClose(true)
+	// set w.closing flag inside draw thread to ensure we can free textures
+	runOnDraw(w, func() {
+		w.viewLock.Lock()
+		w.closing = true
+		w.viewLock.Unlock()
+		w.viewport.SetShouldClose(true)
+		cache.RangeTexturesFor(w.canvas, func(obj fyne.CanvasObject) {
```

Comment:
This is probably not the right place - `Close()` is for programmatic closures, it is not called when user closes the window using the OS drawn accessories - see `closed()` for that.

Reply:
Not sure what you mean here. `closed()` is defined as:
```go
func (w *window) closed(viewport *glfw.Window) {
	viewport.SetShouldClose(false)

	if w.onCloseIntercepted != nil {
		w.QueueEvent(w.onCloseIntercepted)
		return
	}

	w.Close()
}
```
So `Close()` is always executed when `closed` is called unless there is a closeIntercepted callback, however inside that callback (correct me if I am wrong) users should use `Close` if they really want to close the window. At the end, `Close()` is always called 🤔

Reply:
Apologies, I forgot that we changed the order.
However it remains that a window is not truly closed until the driver removes it - `closed` or `Close` just request that it be removed, that may be why you are getting errors in the order of events.

Reply:
However, it may be that I have got my threads crossed with another area. If so apologies and this is fine.

Reply:
> However it remains that a window is not truly closed until the driver removes it - closed or Close just request that it be removed

Yes, that's true, however it should be called here, because to free texture cache effectively we need to delete them using the viewport context and it is only available when the window is not closed, so calling it just before the window is closed seems to be only option.